Humanise @slow

pzrq opened the pull request:
This builds on the discussion in https://github.com/sympy/sympy/pull/9197

Basically the idea is we should be able to make a better trade off when running the majority of the test suite on a local machine for refactoring or development (and have Travis let us know about the really slow final 1%, or alternately use `sympy.test(slow=True)`).

The current trade off is the `@slow` tests on master, after excluding the `ON_TRAVIS` extra-slow tests, take around one hour to run, and the main test suite also takes around one hour to run, i.e. **around 2 hours total** if you really want to know what the Travis output is for at least one Python interpreter before sending it to Travis. 

I think a better trade off would be to make around 1% of those, which happens to be around 49 additional tests, also `@slow`, which reduces the test suite run time to **around 30 minutes**. This branch also finds several tests are probably not worth leaving as `@slow` in the context of their modules where other tests often take >10x longer to run, and removes those `@slow`.

Testing was performed under OSX Python 2.7, Late 2012 Mac Mini, 2.3GHz Intel Core i7, 16GB RAM.

This branch also includes additions to the test framework which print the slowest 3 tests if they took more than 10 seconds each to run, and the symmetric print that if a `@slow` test runs in under 100ms it is considered fast, though others are of course free to determine that something slower than 100ms is fast enough given the context of what is being attempted and that having test coverage by default is more important. 10s and 0.1s (100ms) are based on:
http://www.nngroup.com/articles/response-times-3-important-limits/

Obviously there are exceptions to these, they are hopefully in the commit messages though I would welcome feedback from anyone else in case I've inadvertently made some things `@slow` that really should not be or if there are any other concerns.

My plan is to use this branch whether or not it is ultimately merged in, as I have several other SymPy bugs and regressions I'd like to try to fix, so I expect it to work out as a better trade off for me even if not for anyone else. That said hopefully it is useful :)

pzrq committed:
These tests are no longer slow when compared with other integrals tests.
They take under 1.4 seconds each.

pzrq committed:
Add code that prints up to 3 of the slowest tests if they took more than slow_threshold seconds, currently set at 10.0s based on http://www.nngroup.com/articles/response-times-3-important-limits/

Others can let me know if we should fix the protected vs public TODO, make it configurable instead of 3 tests / 10s threshold, etc.

pzrq committed:
@slow - Add to 49 of the 50 slowest tests that are not already @slow.
Excludes test_code_quality.py:test_files() - seems like a useful one to catch for everyone, as early as possible.
Based on an earlier run that also did not limit the number of slow tests printed to 3, results were as follows.

 tests finished: 5782 passed, 179 skipped, 334 expected to fail, 3 expected to fail but passed, in 3477.55 seconds

test_check_generalized_BAC_CAB_formulas - Took 279.414 seconds
test_prudnikov_2F1 - Took 117.337 seconds
test_gruntz_evaluation - Took 107.473 seconds
test_fidelity - Took 64.412 seconds
test_nonlinear_3eq_order1 - Took 59.793 seconds
test_probability - Took 51.124 seconds
test_lookup_table - Took 37.351 seconds
test_known_facts_consistent - Took 37.260 seconds
test_issue_1796a - Took 35.670 seconds

test_mellin_transform_bessel - Took 31.603 seconds
test_sho - Took 30.938 seconds
test_extract_plane_and_line - Took 30.421 seconds
test_issue_3940 - Took 30.355 seconds
test_evalf_integrals - Took 30.240 seconds
test_properties_of_geometric_objects - Took 30.094 seconds
test_meijerg_expand - Took 29.599 seconds
test_1st_homogeneous_coeff_ode - Took 29.521 seconds
test_inverse_mellin_transform - Took 26.410 seconds
test_meijerg_eval - Took 26.223 seconds
test_plane - Took 25.487 seconds

test_nth_linear_constant_coeff_variation_of_parameters_simplify_False - Took 24.277 seconds
test_nth_linear_constant_coeff_homogeneous_RootOf_sol - Took 24.005 seconds
test_hyperrep - Took 23.653 seconds
test_expint - Took 22.085 seconds
test_gruntz_eval_special - Took 21.010 seconds
test_formulae - Took 20.589 seconds
test_1st_linear - Took 20.430 seconds
test_issue_7130 - Took 19.814 seconds
test_cg_simp_add - Took 19.774 seconds
test_ellipse_geom - Took 19.726 seconds

test_nth_linear_constant_coeff_undetermined_coefficients - Took 19.149 seconds
test_nth_linear_constant_coeff_variation_of_parameters - Took 18.695 seconds
test_mellin_transform - Took 18.652 seconds
test_meijerint - Took 18.636 seconds
test_laplace_transform - Took 18.350 seconds
test_meijerg - Took 17.998 seconds
test_linearize_rolling_disc_kane - Took 17.906 seconds
test_multiple_normal - Took 16.937 seconds
test_files - Took 16.380 seconds
test_random_reduce - Took 16.246 seconds

test_random_insert - Took 16.211 seconds
test_heuristic3 - Took 15.832 seconds
test_dsolve_options - Took 15.828 seconds
test_meijerg_confluence - Took 15.654 seconds
test_lie_group - Took 15.224 seconds
test_nroots1 - Took 15.051 seconds
test_nth_linear_constant_coeff_homogeneous - Took 14.900 seconds
test_uncouple_4_coupled_states - Took 14.247 seconds
test_unrad1 - Took 14.075 seconds
test_issue_4890 - Took 13.591 seconds

test_rsolve - Took 13.560 seconds
test_loggamma - Took 13.433 seconds
test_aux_dep - Took 13.375 seconds
test_inverse_laplace_transform - Took 12.183 seconds
test_expint - Took 12.051 seconds
test_dummy_order_ambiguous - Took 11.932 seconds
test_heuristic1 - Took 11.758 seconds
test_evalf_euler_maclaurin - Took 11.599 seconds
test_gamma_matrix_trace - Took 11.480 seconds
test_issue_2840_8155 - Took 11.407 seconds

test_messy - Took 11.292 seconds
test_free_symbols - Took 11.257 seconds
test_aux - Took 10.762 seconds
test_recursive - Took 10.675 seconds
test_fourier_transform - Took 10.626 seconds
test_polygon - Took 10.600 seconds
test_couple_4_states - Took 10.398 seconds

pzrq committed:
Tests taking under 100ms (0.1s) and so are unambiguously fast on a human timescale.
http://www.nngroup.com/articles/response-times-3-important-limits/

pzrq committed:
Tests that take between 0.1s to 1s aren't slow compared to other tests in their respective modules, so we'd rather take the coverage.

Excludes `test_subrange_large` which seems slow for a Knuth AOCP algorithm, though there's a lot of backtracking and >300k function calls involved so perhaps there are other optimisations possible that @rathmann wished to explore.

pzrq committed:
1.7s is not even close to the 3 minutes test_benchmark_czichowski_buchberger takes.

pzrq committed:
Remove imports from the slow mix, building symbols is known to be relatively slow. On later inspection, 1.4s for a few runs of a matrix algorithm is almost certainly @slow and worth leaving.

pzrq committed:
If test_meijerg_expand - Took 29.599 seconds, before I skipped it, we can surely afford under 2s each for test_prudnikov_9, test_prudnikov_11.

pzrq committed:
Revert marking test_R15 as slow, it used to take ~187s, but if it fails in under 2s no one probably cares (and if they do, just make it @slow again).
Apply same 2s limit pattern to other tests in the module, still leaves around 20 slow tests, many of which are probably better uses of any scarce future development time.

For reference, other unexpectedly fast, i.e. between 2s-10s tests:

Unexpectedly fast: test_bounded, took: 9.660
Unexpectedly fast: test_V13, took: 9.419
Unexpectedly fast: test_unrad_slow, took: 8.691
Unexpectedly fast: test_Y3, took: 6.625
Unexpectedly fast: test_symbolic, took: 6.503
Unexpectedly fast: test_V5, took: 6.282
Unexpectedly fast: test_quintics_2, took: 6.213
Unexpectedly fast: test_V17, took: 6.158
Unexpectedly fast: test_1st_homogeneous_coeff_ode_check134568, took: 6.118

Unexpectedly fast: test_nthroot1, took: 4.907
Unexpectedly fast: test_issue_4511, took: 4.411
Unexpectedly fast: test_H27, took: 3.702
Unexpectedly fast: test_prudnikov_1, took: 3.662
Unexpectedly fast: test_prudnikov_12, took: 3.194
Unexpectedly fast: test_slow_general_univariate, took: 2.854
Unexpectedly fast: test_W23, took: 2.446
Unexpectedly fast: test_H26, took: 2.111

pzrq committed:
For symmetry - the end goal for anyone using this test framework to optimise should be to get it 10x-100x faster, i.e. in most cases under 1s is probably fine on a case by case basis, but a few were under 0.1s and thus definitely could use this idea.

pbrady commented:
@pzrq thanks for taking the time to fix the `@slow` labels.  There was some discussion about adding a pytest-like `--durations` flag which would report test times in #8098.  We decided against it since it's unlikely to be heavily used and we can just use pytest the few times we update testing labels. 

There were only two participants in that discussion so it may be worth hashing things out a bit more here.

toolforger commented:
I'm very much +1 for the overall approach.
I'm also +1 about the changes to import statements (these are generally in a direction I like) and ```@slow``` decorations (I'm pretty sure you have been more systematic than what we have now).

Details can always be improved. Here's the nit I pick:

I'd keep the current values for the time limits, but I'd leave the reference to the webapp time limits out - it may mislead readers into giving these limits credit. And they're essentially arbitrary because the relevant wait time is always the test run, and even a single-test run takes several seconds so we're in the "needs some progress indicator" bracket anyway.
That opens the way to somebody proposing better time limits.

pzrq committed:
Hopefully fix Python 2.6.x - ValueError: zero length field name in format

pzrq committed:
Make fast and slow thresholds configurable, so it is possible for example to test with a lower slow limit than 10 seconds:
sympy.test('sympy/core/tests/test_expr.py', slow_threshold=1)

pzrq committed:
Fix bug where we would always print unexpectedly fast tests, instead of just @slow tests that were unexpectedly fast.

pzrq committed:
Another false positive on unexpectedly fast - should only be unexpectedly fast if we actually ran the slow test. Example usages:

sympy.test('sympy/polys/tests/test_polyroots.py', slow_threshold=1)
...
________________________________ slowest tests _________________________________
test_issue_8438 - Took 1.914 seconds
test_roots0 - Took 2.411 seconds
test_roots_quartic - Took 9.863 seconds
============ tests finished: 18 passed, 1 skipped, in 18.58 seconds ============

sympy.test('sympy/polys/tests/test_polyroots.py', slow_threshold=1, slow=True)
...
________________________________ slowest tests _________________________________
test_nroots1 - Took 14.311 seconds
================== tests finished: 1 passed, in 14.37 seconds ==================

# Hack a return statement into the test_nroots1:
sympy.test('sympy/polys/tests/test_polyroots.py', slow_threshold=1, slow=True)
...
___________________________ unexpectedly fast tests ____________________________
test_nroots1 - Took 0.000 seconds
================== tests finished: 1 passed, in 0.06 seconds ===================

pzrq committed:
We should not need to look up the self._reporter instance all the time, just keep a reference to that instance unless it's another Python 2.6 quirk?

pzrq commented:
@toolforger Just made the slow and fast thresholds configurable and added a comment to hopefully encourage others to come up with a better yardstick if one exists.

To me it's that the test runner doesn't print any single character while it is running a test function, so there is no obvious indication that anything is happening, or if the test has actually hung (i.e. will the [calculation under test even halt](http://en.wikipedia.org/wiki/Halting_problem)?). 

If that's actually because the test is large and can be broken into several smaller tests that give the faster, more granular feedback and remove the `@slow` I'd still count that as an overall win.

pzrq commented:
@pbrady I'm not very familiar with pytest, if `--durations` would be preferred for consistency does it have an equivalent timing flag that would allow future developers to identify tests are now so fast there is no value in leaving the `@slow` decorator on the function? i.e. in an ideal world, the 100 million or so CPU instructions you get per tenth of a second should be enough to do anything worth testing?

On current branch that is something like:

```
In [1]: import sympy
In [2]: sympy.test('sympy/polys/tests/test_polyroots.py', fast_threshold=100, slow_threshold=100, slow=True)

...

sympy/polys/tests/test_polyroots.py[1] .                                    [OK]

___________________________ unexpectedly fast tests ____________________________
test_nroots1 - Took 15.521 seconds
================== tests finished: 1 passed, in 15.59 seconds ==================
Out[2]: True
```

I'm also not sure if `--durations` is on by default? Though that's an easy code fix, I think it's more of a community values or priorities question as to what others think is ultimately important - regressions vs future development speed - as every regression test is ultimately a cost borne by every person who decides to run those tests.

pzrq committed:
Remove limit of 3 - cleaner to filter by sympy.test(slow_threshold=<something bigger than 10>).

pbrady commented:
@pzrq, `pytest --durations N` reports the `N` slowest tests (or if `N=0`, it reports all timings).  Your current changes are more geared towards informing about incorrectly marked tests than reporting test times so I don't think a `--durations` flag would be appropriate here.

pzrq commented:
@rathmann Thanks for the reply, didn't realise Github generated notifications from `@` in commit messages which is pretty cool :)

As you said, `test_subrange_large` is for a reference implementation so it seems appropriate that `@slow` is left on it in case others decide to use it for something else - if I understand correctly they should be using the production version.

Yes the context is PR #9221, basically I went through the test suite looking to see if a better development trade off could be made by putting `@slow` on tests taking significantly longer than 10 seconds, and removing `@slow` from tests which actually aren't very slow (or aren't any more). Almost all tests should still be run on Travis.

Sorry about my sluggish reply, been with family, etc over Easter.

rathmann commented:
Thanks for the reply.   It looks like you are NOT changing the behavior with utilities and iterables/enumerative, is that correct?   

Thanks for doing the change, BTW.  It makes perfect sense that a global perspective will do better than each developer's subjective assessment as to what slow means.

If you do decide to make changes to the tests in utilities,  here is one perspective:

The other half of the equation is what risks the tests guard against. 
utilites/enumerative is almost a leaf -- it depends very little on the rest of Sympy (just compatibility, I think), and is less likely to be broken by a change to a seemingly unrelated part of the codebase.  (The classic regression which unit tests are meant to catch.)   Of course anyone modifying the routine itself should run everything, and hopefully create some new tests besides.

toolforger commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -16,7 +16,7 @@
 from sympy.solvers import solve
 
-from sympy.abc import x, y, z
+from sympy.abc import a, b, c, d, x, y, z
```
This change seems unnecessary...

toolforger commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -2325,14 +2325,12 @@ def test_replace_map():
 
 @slow
 def test_pinv():
-    from sympy.abc import a, b, c, d
```
... it's okay to import locally.
It's not really important either way though. The knee-jerk reaction is "avoid gratuitious change because we want to be able to pinpoint semantic changes in the history", but then imports are mere housekeeping so that particular history isn't very important.

toolforger commented on sympy/utilities/runtests.py:
```diff
@@ -1000,6 +1018,7 @@ def visit_Assert(self, stmt):
         return fix_missing_locations(new_tree)
 
     def test_file(self, filename, sort=True, timeout=False, slow=False, enhance_asserts=False):
+        reporter = self._reporter
```
Is there a reason for this change, other than having a shorthand?
Because usually it's better to stick with the ```self``` prefix: it allows understanding the code without referring back to the assignment. The more the textual length of a scope, the more this applies - here, it's medium-important, and longer ```self._reporter```than ```reporter``` be damned.
(Just my 5 cent, YMMV. Not a reason for rejection.)

toolforger commented:
Sorry for coming back to you so late.

> @toolforger Just made the slow and fast thresholds configurable and
> added a comment to hopefully encourage others to come up with a
> better yardstick if one exists.

I like what I see.

I'm not too happy with measuring wall-clock time, it's too dependent on 
machine performance (and also on CPU load and on whether somebody puts 
the machine in suspended state). However, as a first experimental 
approach this is okay.

I do have some thoughts how to eliminate most relevant sources of error; 
it involves a separate thread that inspects the call stack at regular 
intervals, and measuring time as a percentage of how presences in the 
stack. It would be very different from what you're doing here, but maybe 
more useful across machines.
Whatever the future brings in terms of better performance diagnostics, 
please split this PR in two: One that does the changes around `@slow`, 
and one that adds the thresholds and timing diagnostics. This makes it 
easier to revert the timing diagnostics when/if we want to supersede 
this work.

> To me it's that the test runner doesn't print any single character
> while it is running a test function, so there is no obvious
> indication that anything is happening, or if the test has actually
> hung (i.e. will the [calculation under test even
> halt](http://en.wikipedia.org/wiki/Halting_problem)?).

The tests all finish, except one of them used to eat so many gigabytes 
of memory that it would always overflow memory (I hear that that's fixed 
in the mean time).

> If that's actually because the test is large and can be broken into
> several smaller tests that give the faster, more granular feedback
> and remove the `@slow` I'd still count that as an overall win.

That's marginal, there are factors for and against that.
Usually, however, tests are already at the smallest granularity that the 
original author considered useful, so attacking slowness from this angle 
isn't going to give great rewards in relation to the effort involve.

asmeurer commented:
Printing runtimes is [not difficult](https://github.com/sympy/sympy/commit/a8473c000b7046a6b5eed47b1cd01f554dff2948).

pzrq commented:
Closing for split in #9259, #9260.

@toolforger I hope you succeed in finding a better alternative than wall-clock time or eliminating significant sources of error.

toolforger commented:
@asmeurer not sure what you mean, ```time.time()``` gives wall-clock time.
@pzrq thanks for splitting this. You might want to try ```git checkout <branch> -- <file>``` or ```git commit/stash/stash pop/checkout -b <new branch>``` to transfer changes to a new branch, and keep the original branch.
@ both: ```time.clock()``` should be far less work to get at meaningful CPU cycle times. To get stable-across-machines times, I'd divide by the total running time and express thresholds for ```@slow``` as percentages. Both aspects require verification, so I'll open an Issue for each once this is merged.
